File: scripts/tls-probe.js
const tls = require("tls");

const hosts = [
  { name: "API Backend", host: "api.taskir.com" },
  { name: "Dashboard", host: "dashboard.taskir.com" },
  { name: "Bidding", host: "bidding.taskir.com" },
];

function probeHost({ name, host }) {
  return new Promise((resolve) => {
    const socket = tls.connect(
      {
        host,
        port: 443,
        servername: host,
        minVersion: "TLSv1.3",
        maxVersion: "TLSv1.3",
        rejectUnauthorized: false,
        timeout: 5000,
      },
      () => {
        const cert = socket.getPeerCertificate();
        const result = {
          name,
          host,
          success: true,
          protocol: socket.getProtocol(),
          subject: cert?.subject?.CN || "(unknown)",
          issuer: cert?.issuer?.CN || "(unknown)",
          validFrom: cert?.valid_from || "(unknown)",
          validTo: cert?.valid_to || "(unknown)",
        };
        socket.end();
        resolve(result);
      }
    );

    socket.on("error", (err) => {
      resolve({ name, host, success: false, error: err.message });
    });

    socket.on("timeout", () => {
      socket.destroy();
      resolve({ name, host, success: false, error: "Timeout" });
    });
  });
}

(async () => {
  console.log("TLS 1.3 probe for TaskirX endpoints");
  console.log("======================================");

  for (const entry of hosts) {
    const result = await probeHost(entry);
    if (result.success) {
      console.log(`${result.name}: OK (${result.protocol})`);
      console.log(`  Host     : ${result.host}`);
      console.log(`  Subject  : ${result.subject}`);
      console.log(`  Issuer   : ${result.issuer}`);
      console.log(`  Valid From: ${result.validFrom}`);
      console.log(`  Valid To  : ${result.validTo}`);
    } else {
      console.log(`${result.name}: FAIL`);
      console.log(`  Host : ${result.host}`);
      console.log(`  Error: ${result.error}`);
    }
    console.log("");
  }
})();

File: scripts/tls_probe.go
package main

import (
	"crypto/tls"
	"fmt"
	"net"
	"time"
)

type hostEntry struct {
	name string
	host string
}

func probeTLS13(entry hostEntry) error {
	dialer := &net.Dialer{Timeout: 5 * time.Second}
	config := &tls.Config{
		ServerName: entry.host,
		MinVersion: tls.VersionTLS13,
		MaxVersion: tls.VersionTLS13,
	}
	conn, err := tls.DialWithDialer(dialer, "tcp", fmt.Sprintf("%s:443", entry.host), config)
	if err != nil {
		return err
	}
	defer conn.Close()

	state := conn.ConnectionState()
	cert := state.PeerCertificates

	fmt.Printf("%s: OK (TLS %x)\n", entry.name, state.Version)
	fmt.Printf("  Host     : %s\n", entry.host)
	if len(cert) > 0 {
		fmt.Printf("  Subject  : %s\n", cert[0].Subject.CommonName)
		fmt.Printf("  Issuer   : %s\n", cert[0].Issuer.CommonName)
		fmt.Printf("  NotBefore: %s\n", cert[0].NotBefore.Format(time.RFC3339))
		fmt.Printf("  NotAfter : %s\n", cert[0].NotAfter.Format(time.RFC3339))
	} else {
		fmt.Printf("  Cert     : none\n")
	}
	return nil
}

func main() {
	hosts := []hostEntry{
		{name: "API Backend", host: "api.taskir.com"},
		{name: "Dashboard", host: "dashboard.taskir.com"},
		{name: "Bidding", host: "bidding.taskir.com"},
	}

	fmt.Println("TLS 1.3 probe for TaskirX endpoints")
	fmt.Println("======================================")

	for _, entry := range hosts {
		if err := probeTLS13(entry); err != nil {
			fmt.Printf("%s: FAIL\n", entry.name)
			fmt.Printf("  Host : %s\n", entry.host)
			fmt.Printf("  Error: %s\n\n", err.Error())
			continue
		}
		fmt.Println()
	}
}
